Index: git/include/configs/sun5i.h
===================================================================
--- git.orig/include/configs/sun5i.h
+++ git/include/configs/sun5i.h
@@ -19,6 +19,19 @@
 
 #define CONFIG_SUNXI_USB_PHYS	2
 
+#define CONFIG_BOOTCOMMAND \
+	"if test -n ${fel_booted} && test -n ${scriptaddr}; then " \
+		"echo '(FEL BOOT)';" \
+		"source ${scriptaddr};" \
+	"fi;" \
+	"run mender_setup;" \
+	"mtdparts;" \
+	"ubi part ${mender_mtd_ubi_dev_name};" \
+	"ubifsmount ${mender_uboot_root_name};" \
+	"ubifsload ${fdt_addr_r} /boot/sun5i-r8-chip.dtb;" \
+	"ubifsload ${kernel_addr_r} /boot/zImage;" \
+	"bootz ${kernel_addr_r} - ${fdt_addr_r};"
+
 /*
  * Include common sunxi configuration where most the settings are
  */
@@ -26,4 +39,9 @@
 
 #define CONFIG_MACH_TYPE	(4138 | ((CONFIG_MACH_TYPE_COMPAT_REV) << 28))
 
+/* for mender */
+#define CONFIG_BOOTCOUNT_ENV
+#define CONFIG_BOOTCOUNT_LIMIT
+#define CONFIG_SYS_REDUNDAND_ENVIRONMENT
+
 #endif /* __CONFIG_H */
Index: git/include/configs/sunxi-common.h
===================================================================
--- git.orig/include/configs/sunxi-common.h
+++ git/include/configs/sunxi-common.h
@@ -542,7 +542,8 @@ extern int soft_i2c_gpio_scl;
 	BOOTCMD_SUNXI_COMPAT \
 	"usbnet_devaddr=de:ad:be:af:00:01\0" \
 	"mtdids=nand0=sunxi-nand.0\0" \
-	"mtdparts=mtdparts=sunxi-nand.0:4m(spl),4m(spl-backup),4m(uboot),4m(env),-(UBI)\0" \
+	"mtdparts=mtdparts=sunxi-nand.0:4m(spl),4m(spl-backup),4m(u-boot),8m(u-boot-env),-(ubi)\0" \
+	"bootargs=root=${mender_kernel_root_name} rootfstype=ubifs rw earlyprintk ubi.mtd=${mender_uboot_dev}\0" \
 	BOOTENV
 
 #ifdef CONFIG_CHIPPRO_BOOT_FAST
